tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot

```diff
@@ -0,0 +1,116 @@
+*** Settings ***
+Resource            ../../../../resources/common.resource
+Library             Cumulocity
+Library             ThinEdgeIO
+
+Suite Setup         Custom Setup
+Test Teardown       Get Logs
+
+Test Tags           theme:c8y    theme:troubleshooting    theme:plugins
+
+
+*** Variables ***
+${PARENT_IP}    ${EMPTY}
+${PARENT_SN}    ${EMPTY}
+${CHILD_SN}     ${EMPTY}
+
+
+*** Test Cases ***
+Run custom operation with workflow execution
+    Symlink Should Exist    /etc/tedge/operations/c8y/c8y_TakePicture
+    Cumulocity.Should Contain Supported Operations    c8y_TakePicture
+
+    ${operation}=    Cumulocity.Create Operation
+    ...    description=take a picture
+    ...    fragments={"c8y_TakePicture":{"parameters": {"duration": "5s", "quality": "HD"}}}
+    Verify Local Command    main    take_picture
+    Cumulocity.Operation Should Be SUCCESSFUL    ${operation}
+
+Add template and workflow file dynamically
+    Transfer To Device    ${CURDIR}/c8y_OpenDoor.template    /etc/tedge/operations/c8y/
+    Transfer To Device    ${CURDIR}/open_door.toml    /etc/tedge/operations/
+
+    Should Have MQTT Messages
+    ...    te/device/main///cmd/open_door
+    ...    pattern="^{}$"
+    Symlink Should Exist    /etc/tedge/operations/c8y/c8y_OpenDoor
+    Cumulocity.Should Contain Supported Operations    c8y_OpenDoor
+
+    ${operation}=    Cumulocity.Create Operation
+    ...    description=open a door
+    ...    fragments={"c8y_OpenDoor":{"delay": "5s", "user": "cat"}}
+    Verify Local Command    main    open_door
+    Cumulocity.Operation Should Be SUCCESSFUL    ${operation}
+
+Run custom operation with workflow execution on child device
+    Child Setup
+    Set Test Variable    $CHILD_XID    ${PARENT_SN}:device:${CHILD_SN}
+
+    Set Device Context    ${PARENT_SN}
+    Symlink Should Exist    /etc/tedge/operations/c8y/${CHILD_XID}/c8y_TakePicture
+
+    Cumulocity.Device Should Exist    ${CHILD_XID}
+    Cumulocity.Should Contain Supported Operations    c8y_TakePicture
```

Comment:
question: Is my understanding correct that this line checks if `c8y_TakePicture` is supported by the parent device, not the child device, so it will always be true because we register `c8y_TakePicture` in the suite setup?

I see that we send `114,...,c8y_TakePicture` on MQTT topic `c8y/s/us/TST_augment_connected_sesame:device:TST_bark_randomized_exit`, so we should be checking that it's the child device that supports the operation, right?

Given that both devices support that operation, I don't understand how it ends up on the child device and not the main device.

Reply:
Had a call about this with @Ruadhri17

> Given that both devices support that operation, why does incoming message on c8y/devicecontrol/notifications has externalSource of child device and it ends up being executed on the child device and not the main device?

`Cumulocity.Create Operation` seems to set `externalSource` to the most recent child device that supports the operation, but ultimately which device receives an operation can be decided by the user, using e.g. `go-c8y-cli`. So my concern was just due to my unfamiliarity with this Cumulocity feature and not anything related to thin-edge.

> I see that we send 114,...,c8y_TakePicture on MQTT topic c8y/s/us/TST_augment_connected_sesame:device:TST_bark_randomized_exit, so we should be checking that it's the child device that supports the operation, right?

At the end of the test we verify that the command is executed by the child device, so these previous checks that device exists and device supports the operation may be redundant, but most importantly they are slightly misleading because wrong device is checked.

These checks should be fixed to use a correct device, but they could also just be removed since they're redundant.

Reply:
> Cumulocity.Create Operation seems to set externalSource to the most recent child device that supports the operation, but ultimately which device receives an operation can be decided by the user, using e.g. go-c8y-cli.

Actually, it's the `Cumulocity.Device Should Exist    ${CHILD_XID}` call that changes the context to that device and then all subsequent C8y operations would be done on that device. Yes, implicitly changing the context like that was a bit confusing for me also, when I initially started using this library and have gotten used to it over time. I agree that it would have been better if it was clearer from the API name itself.

Reply:
> Actually, it's the Cumulocity.Device Should Exist ${CHILD_XID} call that changes the context to that device and then all subsequent C8y operations would be done on that device

Ah, that explains it, and as there's no mention of it in the documentation of the keyword, the implicit change of the context is quite unexpected indeed.